Implement ActionInterface for wishlist/shared

dxx004 opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This is a part of the process of introducing implementations of `ActionInterface`.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
N/A

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
N/A

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#29537: Implement ActionInterface for wishlist/shared

Commit:
Implement ActionInterface for Wishlist/Shared

Commit:
Unit test for Wishlist/Shared Controller

Commit:
PSR2 linted

Commit:
Implement ActionInterface for Wishlist/Shared/Cart + unit test

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @dxx004,
Could you review and fix static test failures?
Unit test failure seems to me not related to your changes

ihor-sviziev commented:
Hi @dxx004,
Will you be able to update your PR with requested changes?

ihor-sviziev commented:
@dxx004  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

Commit:
Static test fixes

lbajsarowicz commented:
@ihor-sviziev Reopening due to updates.

Commit:
conflicts resolved; Merge branch '2.4-develop' into refactor/wishlist

# Conflicts:
#	app/code/Magento/Wishlist/Test/Unit/Controller/Shared/AllcartTest.php
#	app/code/Magento/Wishlist/Test/Unit/Controller/Shared/CartTest.php

Commit:
suppress ExcessiveMethodLength notice

ihor-sviziev commented:
Approved and moved to on hold, waiting for architecture decision

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7557](https://jira.corp.magento.com/browse/ENGCOM-7557) has been created to process this Pull Request

ihor-sviziev commented:
FYI this PR was moved to "oh hold" due to internal discussions, reason the same as there https://github.com/magento/magento2/pull/27538#issuecomment-624891873

ihor-sviziev reviewed (changes requested):
Hi @dxx004 ,
We finally have final decision, it clearly described here https://github.com/magento/magento2/pull/27296#pullrequestreview-418541438.

Could you update your PR by bringing back extending abstract action?

ihor-sviziev commented:
@dxx004,
Will you be able to update your PR?

ihor-sviziev commented:
@dxx004  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

lbajsarowicz commented:
As per @dxx004 request, I'm reopening the PR.
What you need to do is to revert `extends` on `Action` class that was inherited previously.
Additionally, Magento Architects decided to keep the backwards-compatible Constructor, so make sure that the elements in Constructor after rebuilding are in the correct order.

https://github.com/dxx004/magento2/blob/f4744395fa245230386c65d00405dfba665dc868/app/code/Magento/Wishlist/Controller/Shared/Allcart.php#L46-L56

Commit:
added extend Action back, updated tests

Commit:
Merge branch '2.4-develop' into refactor/wishlist

ihor-sviziev commented:
@magento run all tests

dxx004 commented:
@ihor-sviziev I see some failed tests, but cannot see why that failed. Could you check pls?

ihor-sviziev commented:
@lenaorobei @slavvka @naydav could you help to identify why builds fail without any report?

lenaorobei commented:
Here is log message:
```
10:13:21 [Build Functional CE] Repositories code generation... 1/8 [==>-----------------]  12% < 1 sec 46.0 MiBPHP Fatal error:  Access level to Magento\Wishlist\Controller\Shared\Allcart::$resultFactory must be protected (as in class Magento\Framework\App\Action\Action) or weaker in /var/www/html/app/code/Magento/Wishlist/Controller/Shared/Allcart.php on line 20 [] []
```

Commit:
removed unnecessary properties

Commit:
Merge branch '2.4-develop' into refactor/wishlist

dxx004 commented:
@magento run all tests

Commit:
fixed tests

Commit:
fixed tests

Commit:
reverted fixed tests

dxx004 commented:
@magento run all tests

Commit:
objectmanager added to test

dxx004 commented:
@magento run all tests

lbajsarowicz commented:
@magento run Functional Tests B2B, Functional Tests EE

ihor-sviziev commented on app/code/Magento/Wishlist/Controller/Shared/Allcart.php:
```diff
@@ -3,55 +3,62 @@
+     * @var ItemCarrier
      */
-    protected $wishlistProvider;
+    private $itemCarrier;
 
     /**
-     * @var \Magento\Wishlist\Model\ItemCarrier
+     * @var WishlistProvider
      */
-    protected $itemCarrier;
+    private $wishlistProvider;
```
Could you revert all not needed changes, changing property visibility to private, changing properties order?

ihor-sviziev commented on app/code/Magento/Wishlist/Controller/Shared/Allcart.php:
```diff
@@ -3,55 +3,62 @@
         Context $context,
-        WishlistProvider $wishlistProvider,
-        ItemCarrier $itemCarrier
+        ItemCarrier $itemCarrier,
+        WishlistProvider $wishlistProvider
```
Please revert this change
```suggestion
        WishlistProvider $wishlistProvider,
        ItemCarrier $itemCarrier
```

ihor-sviziev commented on app/code/Magento/Wishlist/Controller/Shared/Allcart.php:
```diff
@@ -3,55 +3,62 @@
      */
     public function __construct(
         Context $context,
-        WishlistProvider $wishlistProvider,
-        ItemCarrier $itemCarrier
+        ItemCarrier $itemCarrier,
+        WishlistProvider $wishlistProvider
     ) {
-        $this->wishlistProvider = $wishlistProvider;
         $this->itemCarrier = $itemCarrier;
+        $this->wishlistProvider = $wishlistProvider;
```
Please revert this change
```suggestion
        $this->wishlistProvider = $wishlistProvider;
        $this->itemCarrier = $itemCarrier;
```

ihor-sviziev commented on app/code/Magento/Wishlist/Controller/Shared/Cart.php:
```diff
@@ -23,32 +28,32 @@
      * @var CustomerCart
      */
-    protected $cart;
+    private $cart;
```
Please revert changing property visibility for all properties.

ihor-sviziev commented on app/code/Magento/Wishlist/Test/Unit/Controller/Shared/AllcartTest.php:
```diff
@@ -28,19 +28,19 @@ class AllcartTest extends TestCase
-     * @var Context
+     * @var WishlistProvider|MockObject
      */
-    protected $context;
+    protected $wishlistProviderMock;
 
     /**
-     * @var WishlistProvider|MockObject
+     * @var Context|MockObject
      */
-    protected $wishlistProviderMock;
+    protected $context;
```
Please revert changing property position, also would be good to make all property private **in tests**

engcom-Charlie committed:
Merge branch '2.4-develop' into refactor/wishlist

engcom-Charlie committed:
impr

engcom-Charlie commented:
@magento run Unit Tests, Static Tests

engcom-Charlie commented:
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

ihor-sviziev committed:
Merge branch '2.4-develop' into refactor/wishlist

ihor-sviziev commented:
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

engcom-Alfa reviewed (changes requested):
Hi @dxx004 . During testing, we faced the issue.

**Manual testing scenario:**

1. Login to a customer account and add a product to wishlist.
2. Go to "wishlist" and click "Share wishlist".
3. Enter email, message and send.
4. Logout from customer account.
5. Open a shared wishlist email, Click on "View Product" from email.
6. It landed on the shared wishlist products;
 Click the "Add to Cart" button from any product.

**Actual Result:** :heavy_multiplication_x:  404 error message

![Peek 2020-08-11 13-16](https://user-images.githubusercontent.com/51679138/89886862-fc8f7b00-dbd5-11ea-823b-0252f66ca625.gif)

The problem is that it has been implemented `HttpGetActionInterface` instead `HttpPostActionInterface` in `app/code/Magento/Wishlist/Controller/Shared/Allcart.php`

With `HttpPostActionInterface` everything works as expected.
![Peek 2020-08-11 13-28](https://user-images.githubusercontent.com/51679138/89887223-935c3780-dbd6-11ea-8ff3-c6e75532571d.gif)

@ihor-sviziev @dxx004 Could you take a look?

Thanks!

engcom-Charlie committed:
fix http action

engcom-Charlie commented:
@magento run all tests

ihor-sviziev commented:
@engcom-Charlie could you review and fix failing tests?

engcom-Charlie committed:
add get

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@ihor-sviziev could you please review this PR?

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Login to a customer account and add a product to wishlist;
2. Go to "wishlist" and click "Share wishlist";
3. Enter email, message and send;
4. Logout from the customer account;
5. Open a shared wishlist email, Click on "View Product" from email;
6. It landed on the shared wishlist products;
 Click on the "Add to Cart" button from any product.

**Actual Result:** :heavy_check_mark: The product was added to cart

![Screenshot from 2020-08-13 11-02-00](https://user-images.githubusercontent.com/51679138/90110797-75b5dc00-dd56-11ea-848e-6f38f57f12bb.png)

 Click on the "Add All to Cart" button

**Actual Result:** :heavy_check_mark: The products were added to cart

![Screenshot from 2020-08-13 11-01-24](https://user-images.githubusercontent.com/51679138/90110941-9f6f0300-dd56-11ea-9aeb-52c426001675.png)

engcom-Alfa commented:
@magento create issue
